consensus/executor/executor.go

```diff
@@ -0,0 +1,220 @@
+}
+
+// PartialStack contains the ledger features required by the executor.Coordinator
+type PartialStack interface {
+	consensus.LegacyExecutor
+	GetBlockchainInfo() *pb.BlockchainInfo
+}
+
+type coordinatorImpl struct {
+	manager         events.Manager              // Maintains event thread and sends events to the coordinator
+	rawExecutor     PartialStack                // Does the real interaction with the ledger
+	consumer        consensus.ExecutionConsumer // The consumer of this coordinator which receives the callbacks
+	stc             statetransfer.Coordinator   // State transfer instance
+	batchInProgress bool                        // Are we mid execution batch
+	skipInProgress  bool                        // Are we mid state transfer
+}
+
+// NewCoordinatorImpl creates a new executor.Coordinator
+func NewImpl(consumer consensus.ExecutionConsumer, rawExecutor PartialStack, stps statetransfer.PartialStack) consensus.Executor {
+	co := &coordinatorImpl{
+		rawExecutor: rawExecutor,
+		consumer:    consumer,
+		stc:         statetransfer.NewCoordinatorImpl(stps),
+		manager:     events.NewManagerImpl(),
+	}
+	co.manager.SetReceiver(co)
+	return co
+}
+
+// ProcessEvent is the main event loop for the executor.Coordinator
+func (co *coordinatorImpl) ProcessEvent(event events.Event) events.Event {
+	switch et := event.(type) {
+	case executeEvent:
+		if logger.IsEnabledFor(logging.DEBUG) {
+			if logger.IsEnabledFor(logging.DEBUG) {
+				logger.Debug("Executor is processing an executeEvent")
+			}
+		}
+		if co.skipInProgress {
+			logger.Error("FATAL programming error, attempted to execute a transaction during state transfer")
+			return nil
+		}
+
+		if !co.batchInProgress {
+			if logger.IsEnabledFor(logging.DEBUG) {
+				logger.Debug("Starting new transaction batch")
+			}
+			co.batchInProgress = true
+			err := co.rawExecutor.BeginTxBatch(co)
+			_ = err // TODO This should probably panic, see issue 752
+		}
+
+		co.rawExecutor.ExecTxs(co, et.txs)
+
+		co.consumer.Executed(et.tag)
+	case commitEvent:
+		if logger.IsEnabledFor(logging.DEBUG) {
+			logger.Debug("Executor is processing an commitEvent")
+		}
+		if co.skipInProgress {
+			logger.Error("Likely FATAL programming error, attempted to commit a transaction batch during state transfer")
+			return nil
+		}
+
+		if !co.batchInProgress {
+			logger.Error("Likely FATAL programming error, attemted to commit a transaction batch when one does not exist")
+			return nil
+		}
+
+		_, err := co.rawExecutor.CommitTxBatch(co, et.metadata)
+		_ = err // TODO This should probably panic, see issue 752
+
+		co.batchInProgress = false
+
+		info := co.rawExecutor.GetBlockchainInfo()
+
+		co.consumer.Committed(et.tag, info)
+	case rollbackEvent:
+		if logger.IsEnabledFor(logging.DEBUG) {
+			logger.Debug("Executor is processing an rollbackEvent")
+		}
+		if co.skipInProgress {
+			logger.Error("Programming error, attempted to rollback a transaction batch during state transfer")
```

Comment:
I'm on the fence about panic-ing in these situations.  This goes back to #752 about what exactly we should do when we encounter 'irrecoverable type errors', it would be nice if we had a cleaner 'shutdown because of error' than panic.  Anyone else want to weigh in here?